pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -113,15 +113,19 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(env model.Environmen
 			// this is for.
 			FilterChains: make([]plugin.FilterChain, len(l.FilterChains)),
 		}
-		for _, p := range configgen.Plugins {
-			params := &plugin.InputParams{
-				ListenerType:   listenerType,
-				Env:            &env,
-				Node:           &node,
-				ProxyInstances: workloadInstances,
-			}
-			if err = p.OnOutboundListener(params, mutable); err != nil {
-				log.Warna("buildGatewayListeners: failed to build listener for gateway: ", err.Error())
+
+		for _, w := range workloadInstances {
+			for _, p := range configgen.Plugins {
+				params := &plugin.InputParams{
+					ListenerType:    listenerType,
+					Env:             &env,
+					Node:            &node,
+					ProxyInstances:  workloadInstances,
+					ServiceInstance: w,
+				}
+				if err = p.OnOutboundListener(params, mutable); err != nil {
```

Comment:
can you change this to inbound listener uniformly?

Reply:
It's not safe to change the OutboundListener to an InboundListener. OutboundListener is from the current envoy to the mesh, which is what Gateway does here. It seems a bit tricky, but it is correct.